src/Rules/Doctrine/ORM/EntityRule.php

```diff
@@ -0,0 +1,68 @@
+<?php declare(strict_types = 1);
+
+namespace PHPStan\Rules\Doctrine\ORM;
+
+use PhpParser\Node;
+use PHPStan\Analyser\Scope;
+use PHPStan\Rules\Rule;
+use PHPStan\Type\Doctrine\ObjectMetadataResolver;
+use function sprintf;
+
+/**
+ * @implements Rule<Node\Stmt\Class_>
+ */
+class EntityRule implements Rule
+{
+
+	/** @var \PHPStan\Type\Doctrine\ObjectMetadataResolver */
+	private $objectMetadataResolver;
+
+	public function __construct(ObjectMetadataResolver $objectMetadataResolver)
+	{
+		$this->objectMetadataResolver = $objectMetadataResolver;
+	}
+
+	public function getNodeType(): string
+	{
+		return Node\Stmt\Class_::class;
+	}
+
+	public function processNode(Node $node, Scope $scope): array
+	{
+		$objectManager = $this->objectMetadataResolver->getObjectManager();
+		if ($objectManager === null) {
+			return [];
+		}
+
+		$className = (string) $node->namespacedName;
+		try {
+			if ($objectManager->getMetadataFactory()->isTransient($className)) {
+				return [];
+			}
+		} catch (\ReflectionException $e) {
+			return [];
+		}
+
+		try {
+			$metadata = $objectManager->getClassMetadata($className);
+		} catch (\Doctrine\ORM\Mapping\MappingException $e) {
+			return [];
+		}
+
+		$classMetadataInfo = 'Doctrine\ORM\Mapping\ClassMetadataInfo';
+		if (!$metadata instanceof $classMetadataInfo) {
+			return [];
+		}
+
+		$errors = [];
+		if ($node->isFinal()) {
```

Comment:
Ask this first, and only if it's final then continue with checking whether it's an entity etc.